homeassistant/components/mqtt/client.py

```diff
@@ -657,36 +656,29 @@ async def _async_perform_subscriptions(self) -> None:
         # Since we do not know if a published value is retained we need to
         # (re)subscribe, to ensure retained messages are replayed
 
-        def _process_client_subscriptions() -> list[tuple[int, int]]:
-            """Initiate all subscriptions on the MQTT client and return the results."""
-            subscribe_result_list = []
-            for topic, qos in subscriptions.items():
-                result, mid = self._mqttc.subscribe(topic, qos)
-                subscribe_result_list.append((result, mid))
-                _LOGGER.debug("Subscribing to %s, mid: %s, qos: %s", topic, mid, qos)
-            return subscribe_result_list
-
-        subscriptions = self._pending_subscriptions
+        if not self._pending_subscriptions:
+            return
+
+        subscriptions: dict[str, int] = self._pending_subscriptions
         self._pending_subscriptions = {}
 
+        def _process_client_subscriptions() -> tuple[int, int]:
+            """Initiate all subscriptions on the MQTT client and return the results."""
+            result, mid = self._mqttc.subscribe(list(subscriptions.items()))
+            return result, mid
+
         async with self._paho_lock:
-            results = await self.hass.async_add_executor_job(
+            result, mid = await self.hass.async_add_executor_job(
                 _process_client_subscriptions
             )
+        for topic, qos in subscriptions.items():
```

Comment:
Probably still want to move this into the `if result == 0:` block (line 678 below).  It makes no sense to log we're subscribing to something if we got an error, and thus aren't.

Reply:
I do not share that opinion, there's one result, should be clear what we were trying to subscribe at the moment the error was logged.

Reply:
I can understand that.  Though
1. the error does include the message (error_string) when you call_raise_on_error - which should give better info.
2. you now have a problem of _n_ Subscribe log messages to 1 error message, and no idea which of those subscribe log messages the error is related to (if you have two rounds of subscribing, and the first succeeded and second didn't, you won't be able to tell which subscribed and which errored),
3. If you got an error, you're NOT subscribing to those topics (so the log message is wrong about what's happening).  If you get an error, it couldn't issue the SUBSCRIBE - which is why it's not waiting on the MID (which is the response from the server).

But this is not a hill I'm going to die on, this should fail very infrequently anyway.

Possible alternative - in the result == 0 path move 'subscribing to' and in the else add a 'failed to subscribe to' log message?  Assuming the mqtt.error_string doesn't already cover it.

Reply:
The logging is for debugging purposes, there is no problem. If the connection fails, the broker will be disconnected and the ha subscriptions will be resubscribed. The part below is just for throwing or awaiting the ACK.